pkg/alicloud/client/client_dns.go

```diff
@@ -21,17 +21,30 @@ import (
 	"time"
 
 	"github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud"
+	"golang.org/x/time/rate"
 
 	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/errors"
 	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
 	"github.com/aliyun/alibaba-cloud-sdk-go/services/alidns"
 	"github.com/gardener/gardener/pkg/utils"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
 const (
-	domainsCacheTTL = 1 * time.Hour
+	domainsCacheTTL     = 1 * time.Hour
+	rateLimiterCacheTTL = 1 * time.Hour
 )
 
+// RateLimiterWaitError is an error to be reported if waiting for a aliyun dns  rate limiter fails.
+// This can only happen if the wait time would exceed the configured wait timeout.
+type RateLimiterWaitError struct {
+	Cause error
+}
+
+func (e *RateLimiterWaitError) Error() string {
+	return fmt.Sprintf("could not wait for client-side route53 rate limiter: %+v", e.Cause)
+}
+
 // NewDNSClient creates a new DNS client with given region, accessKeyID, and accessKeySecret.
 func (f *clientFactory) NewDNSClient(region, accessKeyID, accessKeySecret string) (DNS, error) {
 	client, err := alidns.NewClientWithAccessKey(region, accessKeyID, accessKeySecret)
```

Comment:
How about cache the client instead? Aliyun client is threadsafe.

Reply:
Currently both domain cache and ratelimiter cache are on factory level which benefits all dnsclient.  It's not so necessary to add another cache level.  From other side, if we cache dnsclient, the Get method of ratelimiter is never called anymore and TTL won't be refreshed.
So I don't suggest to do so now